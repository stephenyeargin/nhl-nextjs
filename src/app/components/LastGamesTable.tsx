import React from 'react';
import Link from 'next/link';
import TeamLogo from '@/app/components/TeamLogo';
import { formatLocalizedDate, formatStat } from '@/app/utils/formatters';

type StatHeader = {
  key: string;
  label: string;
  title: string;
  altKey?: string;
  precision?: number;
  unit?: string;
};

export interface PlayerGameSummary {
  gameId: string | number;
  gameDate: string;
  homeRoadFlag: 'H' | 'A';
  teamAbbrev: string;
  opponentAbbrev: string;
  [k: string]: unknown;
}

interface LastGamesTableProps {
  games: PlayerGameSummary[];
  statHeaders: readonly StatHeader[];
  headerColorClass?: string;
  headerStyle?: React.CSSProperties;
}

const LastGamesTable: React.FC<LastGamesTableProps> = ({
  games,
  statHeaders,
  headerColorClass,
  headerStyle,
}) => {
  const visibleHeaders = statHeaders.filter((h) => {
    const key = h.key;
    const altKey = h.altKey;
    const present = games.some(
      (g) =>
        Object.prototype.hasOwnProperty.call(g, key) ||
        (altKey && Object.prototype.hasOwnProperty.call(g, altKey))
    );

    return present;
  });

  return (
    <div className="overflow-x-auto">
      <table className="text-sm w-full">
        <thead>
          <tr className={`${headerColorClass ?? ''}`} style={headerStyle}>
            <th className="p-2 text-center w-10" style={headerStyle}>
              Date
            </th>
            <th className="p-2 text-left" style={headerStyle}>
              Opponent
            </th>
            {visibleHeaders.map((h) => (
              <th key={h.key} className="p-2 text-center" style={headerStyle}>
                <abbr className="underline decoration-dashed" title={h.title}>
                  {h.label}
                </abbr>
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {games.map((g, i) => (
            <tr key={i} className={`${i % 2 ? 'bg-slate-500/10' : ''}`}>
              <td className="p-2 border text-center">{formatLocalizedDate(g.gameDate)}</td>
              <td className="p-2 border text-left">
                <div className="font-bold">
                  {g.homeRoadFlag !== 'H' ? (
                    <div className="flex items-center gap-2">
                      <TeamLogo
                        team={g.teamAbbrev}
                        className="hidden md:block h-8 w-8"
                        alt={g.teamAbbrev}
                      />
                      <Link href={`/game/${g.gameId}`} className="font-bold underline">
                        {g.teamAbbrev}@{g.opponentAbbrev}
                      </Link>
                      <TeamLogo
                        team={g.opponentAbbrev}
                        className="hidden md:block h-8 w-8"
                        alt={g.opponentAbbrev}
                      />
                    </div>
                  ) : (
                    <div className="flex items-center gap-2">
                      <TeamLogo
                        team={g.opponentAbbrev}
                        className="hidden md:block h-8 w-8"
                        alt={g.opponentAbbrev}
                      />
                      <Link href={`/game/${g.gameId}`} className="font-bold underline">
                        {g.opponentAbbrev}@{g.teamAbbrev}
                      </Link>
                      <TeamLogo
                        team={g.teamAbbrev}
                        className="hidden md:block h-8 w-8"
                        alt={g.teamAbbrev}
                      />
                    </div>
                  )}
                </div>
              </td>
              {visibleHeaders.map((h) => {
                const key = h.key;
                const altKey = h.altKey;
                const precision = h.precision;
                const unit = h.unit;
                const primary = g[key as keyof typeof g];
                const alt = altKey ? (g[altKey as keyof typeof g] as any) : undefined;
                const val =
                  typeof primary === 'number' || typeof primary === 'string'
                    ? (primary as any)
                    : typeof alt === 'number' || typeof alt === 'string'
                      ? (alt as any)
                      : undefined;

                if (val === undefined) {
                  return (
                    <td key={key} className="p-2 border text-center text-xs">
                      --
                    </td>
                  );
                }

                if (key === 'plusMinus') {
                  const n = typeof val === 'number' ? val : Number(val);
                  const content = Number.isNaN(n) ? '--' : n > 0 ? `+${n}` : `${n}`;

                  return (
                    <td key={key} className="p-2 border text-center text-xs">
                      {content}
                    </td>
                  );
                }

                return (
                  <td key={key} className="p-2 border text-center text-xs">
                    {formatStat(val as any, precision, unit)}
                  </td>
                );
              })}
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
};

export default LastGamesTable;
