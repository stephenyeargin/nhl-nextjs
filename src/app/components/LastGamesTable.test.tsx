import React from 'react';
import { render, screen, within } from '@testing-library/react';
import LastGamesTable, { PlayerGameSummary } from '@/app/components/LastGamesTable';

const headers = [
  { key: 'goals', label: 'G', title: 'Goals' },
  { key: 'plusMinus', label: '+/-', title: 'Plus/Minus' },
] as const;

const mkGame = (overrides: Partial<PlayerGameSummary>): PlayerGameSummary => ({
  gameId: '2024010101',
  gameDate: '2024-01-01',
  homeRoadFlag: 'A',
  teamAbbrev: 'NSH',
  opponentAbbrev: 'COL',
  ...overrides,
});

describe('LastGamesTable', () => {
  it('renders dynamic headers and placeholder cells', () => {
    const games: PlayerGameSummary[] = [
      mkGame({ goals: 2, plusMinus: 1 }),
      mkGame({ goals: undefined as any }),
    ];

    render(<LastGamesTable games={games} statHeaders={headers} />);

    // Headers
    expect(screen.getByText('G')).toBeInTheDocument();
    expect(screen.getByText('+/-')).toBeInTheDocument();

    const rows = screen.getAllByRole('row');
    const r1 = rows[1];
    expect(within(r1).getByText('2')).toBeInTheDocument();
    expect(within(r1).getByText('+1')).toBeInTheDocument();

    const r2 = rows[2];
    // Should render '--' where value is missing
    expect(within(r2).getAllByText('--').length).toBeGreaterThan(0);
  });

  it('matches snapshot (two games)', () => {
    const games: PlayerGameSummary[] = [
      mkGame({ goals: 1, plusMinus: -1 }),
      mkGame({ goals: 0, plusMinus: 2 }),
    ];

    const { container } = render(<LastGamesTable games={games} statHeaders={headers} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
