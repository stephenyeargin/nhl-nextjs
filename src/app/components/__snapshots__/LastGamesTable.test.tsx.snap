// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LastGamesTable matches snapshot (two games) 1`] = `
<div
  class="overflow-x-auto"
>
  <table
    class="text-sm w-full"
  >
    <thead>
      <tr
        class=""
      >
        <th
          class="p-2 text-center w-10"
        >
          Date
        </th>
        <th
          class="p-2 text-left"
        >
          Opponent
        </th>
        <th
          class="p-2 text-center"
        >
          <abbr
            class="underline decoration-dashed"
            title="Goals"
          >
            G
          </abbr>
        </th>
        <th
          class="p-2 text-center"
        >
          <abbr
            class="underline decoration-dashed"
            title="Plus/Minus"
          >
            +/-
          </abbr>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class=""
      >
        <td
          class="p-2 border text-center"
        >
          1/1/2024
        </td>
        <td
          class="p-2 border text-left"
        >
          <div
            class="font-bold"
          >
            <div
              class="flex items-center gap-2"
            >
              <a
                href="/team/NSH"
              >
                <img
                  alt="NSH"
                  class="hidden md:block h-8 w-8"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="https://assets.nhle.com/logos/nhl/svg/NSH_light.svg"
                  style="color: transparent;"
                  width="256"
                />
              </a>
              <a
                class="font-bold underline"
                href="/game/2024010101"
              >
                NSH
                @
                COL
              </a>
              <a
                href="/team/COL"
              >
                <img
                  alt="COL"
                  class="hidden md:block h-8 w-8"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="https://assets.nhle.com/logos/nhl/svg/COL_light.svg"
                  style="color: transparent;"
                  width="256"
                />
              </a>
            </div>
          </div>
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          1
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          -1
        </td>
      </tr>
      <tr
        class="bg-slate-500/10"
      >
        <td
          class="p-2 border text-center"
        >
          1/1/2024
        </td>
        <td
          class="p-2 border text-left"
        >
          <div
            class="font-bold"
          >
            <div
              class="flex items-center gap-2"
            >
              <a
                href="/team/NSH"
              >
                <img
                  alt="NSH"
                  class="hidden md:block h-8 w-8"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="https://assets.nhle.com/logos/nhl/svg/NSH_light.svg"
                  style="color: transparent;"
                  width="256"
                />
              </a>
              <a
                class="font-bold underline"
                href="/game/2024010101"
              >
                NSH
                @
                COL
              </a>
              <a
                href="/team/COL"
              >
                <img
                  alt="COL"
                  class="hidden md:block h-8 w-8"
                  data-nimg="1"
                  decoding="async"
                  height="256"
                  loading="lazy"
                  src="https://assets.nhle.com/logos/nhl/svg/COL_light.svg"
                  style="color: transparent;"
                  width="256"
                />
              </a>
            </div>
          </div>
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          0
        </td>
        <td
          class="p-2 border text-center text-xs"
        >
          +2
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
